castervoice/core/context.py

```diff
@@ -0,0 +1,45 @@
+from dragonfly.grammar.context import LogicAndContext
+from dragonfly import (
+        AppContext,
+        Context as DragonflyContext
+        )
+
+
+class Context(DragonflyContext):
```

Comment:
Currently the support only dragonfly `AppContext` other context classes such as FuncContext?
In addition another limitation is [logical operations](https://dragonfly2.readthedocs.io/en/latest/context.html?highlight=AppContext#logical-operations) on those context.

Historically and caster we were not limited but it was a design decision which I regret moving to the grammar registry system to which is being re implemented here.

Reply:
This is kind of a Caster-type context which is represented as a context from the config.

So, one of this `Context` object is instantiated per context in the config, for example: https://github.com/Timoses/caster-core/blob/a4a5370c14f0346da2c97bf87cd3b5f228742adf/config/caster.yml#L41-L44


From the config file it is also obvious that I could define something like this in the config:
```
- name: my_firefox_context
  executable: firefox
  plugins:
    - firefox_plugin
```

So above would be a context which is active when the executable is `firefox`.


The following context definition would extend the firefox context and only be active when the plugin `firefox` would be in the context `active_field: search_bar` (this is a context provided by the plugin. Plugins can freely define what kind of values and fields can be used to define a desired context). And in the context the `firefox_searcher` plugin would be enabled (i.e. the plugin's grammar) :
```
- name: my_firefox_searchbar_context
  extends: my_firefox_context
  firefox_plugin:
    active_field: search_bar
  plugins:
    - firefox_searcher
```

So the plugin provided contexts can of course also be FuncContexts (or any Dragonfly contexts).


> Historically and caster we were not limited but it was a design decision which I regret moving to the grammar registry system to which is being re implemented here.

What else would you envision? Perhaps create an issue for discussion or point me to one if one already exists.

Reply:
In effect we provide the user the means to define her or his contexts via the configuration file.

We however can use whatever backend implementation for contexts that we desire.


Of course both sides of "frontend" (user-facing config file (/later perhaps GUI)) and "backend" (implementation of these user-defined contexts) may be up for discussion. IMO the focus should be on the "frontend" usability.

Reply:
> So the plugin provided contexts can of course also be FuncContexts (or any Dragonfly contexts).

>    Historically and caster we were not limited but it was a design decision which I regret moving to the grammar registry system to which is being re implemented here.

> What else would you envision? Perhaps create an issue for discussion or point me to one if one already exists. 

I wouldn't envision anything different as it's possible to Not to fall in the same design pattern with a grammar registry system. It wasn't clear to me how this worked out in the code with dragonfly. As I become more familiar with the codeand the documentation this will become clear.

Reply:
> IMO the focus should be on the "frontend" usability.

I agree with this.